src/core/include/openvino/core/type.hpp

```diff
@@ -77,9 +77,18 @@ struct OPENVINO_API DiscreteTypeInfo {
 OPENVINO_API
 std::ostream& operator<<(std::ostream& s, const DiscreteTypeInfo& info);
 
+namespace frontend {
+class ConversionExtensionBase;
+} // frontend
+
+template <typename T>
+constexpr bool use_ov_dynamic_cast() {
 #if defined(__ANDROID__) || defined(ANDROID)
-#    define OPENVINO_DYNAMIC_CAST
+    return true;
+#else
+    return std::is_base_of_v<ov::frontend::ConversionExtensionBase, T>;
```

Comment:
I find it a bit odd that we have common code that depends on some specific details. (ov::frontend::ConversionExtensionBase).

I'd suggest to add special conversion func for ConversionExtensions. And use it from FEs instead of as_type_ptr.
Smth like that (i didn't test it):
template <typename Type, typename Value, typename = std::enable_if_t<!std::is_base_of_v<ConversionExtensionBase, Type>>>
auto conversion_extension_as_type_ptr(const Value& value) -> decltype(ov::as_type_ptr<Type>(value)) {
    return ov::as_type_ptr<Type>(value);
}

But if everyone else is ok with it, then OK :)

Reply:
I agree with the conclusion. As of solution I'd apply some type traits based on integral_const.